File: Top Interview Questions/L102二叉树的层序遍历.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
const levelOrder = function(root) {
    let ans=[]
    let queue=[]
    if(root==null){
        return ans
    }
    queue.push(root)
    process(queue,ans)
    return ans
};
function process(queue,ans){
    if(queue.length==0){
        return
    }

    let hp=queue.length
    let res=[]
    for(let i=0;i<hp;i++){
        let p=queue.shift()
        if(p){
            res.push(p.val===undefined?0:p.val)
            if(p.left!==null) queue.push(p.left)
            if(p.right!==null) queue.push(p.right)
        }
    }
    ans.push(res)

    process(queue,ans)
}

File: java-devildyw/src/main/java/top/dyw/leetcode/medium/tree/L102二叉树的层序遍历.java
package top.dyw.leetcode.medium.tree;


import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Devil
 * @since 2022-07-21-11:54
 * <p>
 * 给你二叉树的根节点 root ，返回其节点值的 层序遍历 。 （即逐层地，从左到右访问所有节点）。
 */
@SuppressWarnings("all")
public class L102二叉树的层序遍历 {
    public static void main(String[] args) {

    }

    /**
     * 使用队列
     *
     * @param root
     * @return
     */
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> arrayLists = new ArrayList<>();
        Deque<TreeNode> deque = new LinkedList<>();
        if (root != null) {
            deque.offer(root);
        }
        while (!deque.isEmpty()) {
            int size = deque.size();
            ArrayList<Integer> list = new ArrayList<>();
            while (size-- > 0) {
                TreeNode poll = deque.poll();
                list.add(poll.val);
                if (poll.left != null) {
                    deque.offer(poll.left);
                }
                if (poll.right != null) {
                    deque.offer(poll.right);
                }
            }
            arrayLists.add(list);
        }
        return arrayLists;
    }
}
